src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+            {
+               wxLogError("Failed to read bus info");
+               return 0;
+            }
+            if(busInfo.busType == Vst::kMain)
+            {
+               data.outputs[busIndex].numChannels = busInfo.channelCount;
+               data.outputs[busIndex].channelBuffers32 = outBlock + outputBlocksOffset;
+               outputBlocksOffset += busInfo.channelCount;
+            }
+            else
+            {
+               //aux is not yet supported
+               data.outputs[busIndex].numChannels = 0;
+               data.outputs[busIndex].channelBuffers32 = nullptr;
+            }
+            data.outputs[busIndex].silenceFlags = 0UL;
+         }
+      }
+      
+      const auto processResult = audioProcessor->process(data);
+      
+      return processResult == kResultOk ?
+         data.numSamples : 0;
+   }
+   wxLogError("Failed to query IAudioProcess interface from the effect component");
+   return 0;
+   }
+}
+
+size_t VST3Effect::ProcessBlock(float** inBlock, float** outBlock, size_t blockLen)
+{
+   internal::ComponentHandler::PendingChangesPtr pendingChanges { nullptr };
+   if(mComponentHandler)
+      pendingChanges = mComponentHandler->getPendingChanges();
+   return VST3ProcessBlock(mEffectComponent.get(), mSetup, inBlock, outBlock, blockLen, pendingChanges.get());
+}
+
+bool VST3Effect::RealtimeInitialize()
+{
+   //reload current parameters form the editor into parameter queues
+   SyncParameters();
+   return true;
+}
+
+bool VST3Effect::RealtimeAddProcessor(unsigned numChannels, float sampleRate)
+{
+   using namespace Steinberg;
+
+   auto effect = std::make_unique<VST3Effect>(*this);
+   effect->mSetup.processMode = Vst::kRealtime;
+   effect->mSetup.sampleRate = sampleRate;
+   
+   if(!effect->ProcessInitialize({0}, nullptr))
+   {
+      wxLogError("VST3 realtime initialization failed");
+      return false;
+   }
+
+   mRealtimeGroupProcessors.push_back(std::move(effect));
+   return true;
+}
+
+bool VST3Effect::RealtimeFinalize()
+{
+   for(auto& processor : mRealtimeGroupProcessors)
+      processor->ProcessFinalize();
+
+   mRealtimeGroupProcessors.clear();
+   mPendingChanges.reset();
+   
+   return true;
+}
+
+bool VST3Effect::RealtimeSuspend()
+{
+   using namespace Steinberg;
+   for(auto& effect : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(effect->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, reinterpret_cast<void**>(&audioProcessor)) == kResultOk)
+         audioProcessor->setProcessing(false);
+   }
+   return true;
+}
+
+bool VST3Effect::RealtimeResume()
+{
+   using namespace Steinberg;
+   for(auto& effect : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(effect->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, reinterpret_cast<void**>(&audioProcessor)) == kResultOk)
+         audioProcessor->setProcessing(true);
```

Comment:
I see suspend and resume only call `setProcessing` for slaves, not for `this`.  That is not like VST 2.  However maybe VST 2 did that unnecessarily, because it was only for the unnecessary calculations that I removed at commit 81951abe5dbc3c263bc5e4d40ccb3e685bfa0170.  Is this what you were thinking?  Then maybe my commit was not complete, in its removal of unnecessary things.

Reply:
Maybe VST2 implementation does not need to initialize the "master effect" too, there is no need to do so in the VST3 implementation since all the processing is done by "processor effect" instances, in the case of realtime processing  master effect simply holds initial parameters with which other instances would be initialized